stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2005 @@
+            stacks_client
+                .submit_block_for_validation(block.clone())
+                .unwrap_or_else(|e| {
+                    warn!(
+                        "Signer #{}: Failed to submit block for validation: {e:?}",
+                        self.signer_id
+                    );
+                });
+        }
+    }
+
+    /// Process inbound packets as both a signer and a coordinator
+    /// Will send outbound packets and operation results as appropriate
+    fn handle_packets(
+        &mut self,
+        stacks_client: &StacksClient,
+        res: Sender<Vec<OperationResult>>,
+        packets: &[Packet],
+        current_reward_cycle: u64,
+    ) {
+        let signer_outbound_messages = self
+            .signing_round
+            .process_inbound_messages(packets)
+            .unwrap_or_else(|e| {
+                error!(
+                    "Signer #{}: Failed to process inbound messages as a signer: {e:?}",
+                    self.signer_id
+                );
+                vec![]
+            });
+
+        // Next process the message as the coordinator
+        let (coordinator_outbound_messages, operation_results) = self
+            .coordinator
+            .process_inbound_messages(packets)
+            .unwrap_or_else(|e| {
+                error!(
+                    "Signer #{}: Failed to process inbound messages as a coordinator: {e:?}",
+                    self.signer_id
+                );
+                (vec![], vec![])
+            });
+
+        if !operation_results.is_empty() {
+            // We have finished a signing or DKG round, either successfully or due to error.
+            // Regardless of the why, update our state to Idle as we should not expect the operation to continue.
+            self.process_operation_results(stacks_client, &operation_results, current_reward_cycle);
+            self.send_operation_results(res, operation_results);
+            self.finish_operation();
+        } else if !packets.is_empty() && self.coordinator.state != CoordinatorState::Idle {
+            // We have received a message and are in the middle of an operation. Update our state accordingly
+            self.update_operation();
+        }
+        self.send_outbound_messages(signer_outbound_messages);
+        self.send_outbound_messages(coordinator_outbound_messages);
+    }
+
+    /// Validate a signature share request, updating its message where appropriate.
+    /// If the request is for a block it has already agreed to sign, it will overwrite the message with the agreed upon value
+    /// Returns whether the request is valid or not.
+    fn validate_signature_share_request(&self, request: &mut SignatureShareRequest) -> bool {
+        let message_len = request.message.len();
+        // Note that the message must always be either 32 bytes (the block hash) or 33 bytes (block hash + b'n')
+        let hash_bytes = if message_len == 33 && request.message[32] == b'n' {
```

Comment:
Not necessary in this PR, but there's an inner message protocol at work here that should be better encapsulated.  This really should be captured into its own struct which hides the details of how to get the hash bytes and how to interpret the byte prefix.